### ISSUE TITLE
Assert macro conflicts

### ISSUE THREAD

**Issue by Snaipe** [Snaipe/Criterion#20]
the `assert` macro conflicts with the one provided by the standard `assert.h`. This can be a problem when both are included in any order.

* Should I provide an alias for `assert` ? (maybe `criterion_assert`)
* Should I undefine the assert from `assert.h` and replace it with mine ?

also, in the event that criterion.h is included before assert.h, it should define `_ASSERT_H` to let the posix header replace assert itself (The MSVC header seem to replace it regardless)

**Comment by bturrubiates** [comment]
This is an issue I've been encountering recently. Current temporary fix is to `#undef assert` after including something that has included `assert.h` and include the `criterion.h` headers last. Personally, I don't think I'd like it if a library internally did an `#undef assert` and defined `_ASSERT_H`. Consider a shorter alias such as `c_assert`. Doesn't look quite as nice as using `assert`, but it's an easier solution that also makes it easier to distinguish the call.

**Comment by Snaipe** [comment]
@bturrubiates would you rather completely replace `assert` with something like `c_assert`, or just not redefine `assert` when the macro is already defined and define `c_assert` instead (with maybe a compile-time notice) ?

**Comment by bturrubiates** [comment]
Well personally I think I'd rather see it always be defined as a non-conflicting name. If you do end up redefining `assert` and `_ASSERT_H` I think it's appropriate to present a compile-time notice, but also provide some information in the docs about how Criterion should be included last to avoid any issues.

**Comment by mkfifo** [comment]
Just thought I would add my 2 cents as a random bystander; I would much prefer for Criterion to not conflict with a c lib name.

Typing a few extra character is not a problem `c_assert,` `cassert`, etc. are all fine.

**Comment by Snaipe** [comment]
This seems reasonable.
The `c_` prefix seems to be reserved by `termios.h`, so I will probably go with `cr_` (and the prefix is not reserved by the standard library or posix) [[source]](http://www.gnu.org/software/libc/manual/html_node/Reserved-Names.html).
For consistency, I will most likely need to rename all the other macros to add the new prefix -- this will break backward compatibility, so I'll add aliases on the old macros along with a compiler deprecation notice.

**Comment by Snaipe** [comment]
I added the prefix and the deprecation message.

Old aliases may be disabled by defining the `CRITERION_NO_COMPAT` macro, but it shouldn't be necessary (and those aliases will be removed on the next major version).

The only special case I ended up doing is for `assert`; I still defined the macro if it does not exist, along with `_ASSERT_H`. This has the following effects:
* if `criterion.h` is included alone, `assert` is defined as an alias to `cr_assert` with a deprecation message.
* if `assert.h` is included before `criterion.h`, the macro does not get redefined by criterion.
* if `assert.h` is included after `criterion.h`, the standard library header replaces `assert` with its own macro.

All in all, if `assert.h` gets included somewhere, then `assert` is the standard library one; otherwise, the user gets a deprecation message. This seems like the best solution to address the backward-compatibility problem for `assert`.
